Cross-compile Instruction for Linux targeting Windows

dnabre opened the pull request:
Instructions on setting up cross-build environment, and cross-compiling htop for Windows. 

Initial setup in providing a native port to Windows.  See Issue #1097

dnabre commented:
Clean up and previous accidental PR. 
Markdown instructions (reformatted and generally edited) along with expository scripts

BenBE commented on docs/winnt/Build.Cross-compile with minGW64 for Windows.md:
```diff
@@ -0,0 +1,238 @@
+            libltdl-dev libssl-dev libtool-bin      \\
+            libxml-parser-perl                      \\
+            make openssl                            \\
+            patch perl python3 ruby                 \\
```
Which part of this needs Ruby? What about the Perl and Python dependencies - where are they coming from?

dnabre replied:
I pulled the prerequisites from MXE, there were what was needed to compile a whole pile of things from source for cross-compiling. Scrapped the list entirely, and worked from clean install + build essentials. Also checked if anything different would be need on Ubuntu (since GH CI stuff is setup on Ubuntu), and documented additions.

BenBE commented on docs/winnt/Build.Cross-compile with minGW64 for Windows.md:
````diff
@@ -0,0 +1,238 @@
+
+```
+     apt install \\
+            autoconf automake autopoint             \\
+            bash bison build-essential bzip2        \\
+            flex                                    \\
+            g++ g++-multilib gettext git gperf      \\
+            intltool                                \\
+            libc6-dev-i386 libgdk-pixbuf2.0-dev     \\ 
+            libltdl-dev libssl-dev libtool-bin      \\
+            libxml-parser-perl                      \\
+            make openssl                            \\
+            patch perl python3 ruby                 \\
+            sed sudo                                \\
+            unzip zip  xz-utils                     \\
+            wget
````
Please only list explicit dependencies here. Anything non-obvious like e.g. `libgdk-pixbuf2.0-dev` should get a short mention for why it is included.

dnabre replied:
Fixed as noted above. God knows where libgdk-pixbuf came from

BenBE commented on docs/winnt/Build.Cross-compile with minGW64 for Windows.md:
````diff
@@ -0,0 +1,238 @@
+
+For a solid build environment, we want to install these packages(we'll include tools for both 32-bit and 64-bit targets). This list definitely isn't minimal, but given a solid base to compiler most anything, including any additional compiling cross-compilers if they ever are needed. 
+
+```
````
No need for the additional indentation inside code blocks.

dnabre replied:
Extra indentations, blank lines, and whitespace in code blocks removed.

BenBE commented on docs/winnt/Build.Cross-compile with minGW64 for Windows.md:
```diff
@@ -0,0 +1,238 @@
+
+For the required cross-compilation, we just need the one meta-package:
+
+    apt install mingw-w64
```
Should use code blocks with back ticks.

dnabre replied:
fixed

BenBE commented on docs/winnt/build_ncurses.sh:
```diff
@@ -0,0 +1,47 @@
+	--with-fallbacks=ms-terminal \
+	--without-manpages
+
+# standard 'make' and 'make install' from here
```
Missing newline at EOF.

dnabre commented:
Thanks for the detailed feedback.

Working through the fixes to document, what needs changed/fixed is clear. No problem with fixing the document itself.

I'm not experienced with a lot parts of  github (solo development only exposes you to so much) . I'm not sure how to properly fix and resolve each individual conversation. I'm looking up docs and going to Discord to ask questions. Please pardon any improper PR, commits, or what I might spawn.

BenBE commented:
Don't worry about the commit cleanup – I can support with that or take care of the final polishing. If you need help with git feel free to ask. Ideally you can rebase your commits (cf. `git commit --fixup=<hash>` --> `git rebase -i --autosquash main` --> git push --force-with-lease <your-repo> <your branch>`) on this branch; they will directly show up here as changed.

If you think one of the items of the conversations is done, just drop a note in there.

---
BTW: You might like https://ohmygit.org/ for a playful way to learn working with git.

dnabre commented on docs/winnt/Build.Cross-compile with minGW64 for Windows.md:
```diff
@@ -0,0 +1,238 @@
+---
+ 
+ 
+ Work in progress. See Issue #1097 for details.  Initial work is being done by @dnabre , starting 2022/10/15.
```
Unless I'm missing something, just switching from '/' to '-' was all need for ISO 8601 (example 2022-10-15). fixed

Never heard the term of plenk before, but apparently I tend to do it on after tokens I except to be parsed. Didn't find too many of these places,  but tried to check all of them.

dnabre commented on docs/winnt/Build.Cross-compile with minGW64 for Windows.md:
```diff
@@ -0,0 +1,238 @@
+
+* **winnt**: Name of the Windows kernel. This name is used internally to designed code for the Windows port. The constant `HTOP_WINNT` when building for Windows.  
+
+* **i686** and **x86_64**:  These are terms of the common Intel-based 32-bit and 64-bit cpu architectures in common use.  The latter, `x86_64` is more common by far nowadays, so for brevity I'll use it, but generally it could be replaced with `i686` wherever used for its 32-bit counterpart.   
```
Did pass through, remove all manual lines break. 

Literals all back ticked, and checked for consistency. 

Abbreviations capitalized, example CPU ended up being removed. Capitalization consistency  checked and fixed.

dnabre commented on docs/winnt/Build.Cross-compile with minGW64 for Windows.md:
```diff
@@ -0,0 +1,238 @@
+
+---
+
+### Fun naming conventions bits:
```
With horizontal line removal, redid all heading/subheadings.

dnabre commented on docs/winnt/Build.Cross-compile with minGW64 for Windows.md:
```diff
@@ -0,0 +1,238 @@
+# Cross-compiling htop for Windows from Linux
```
When through a bunch of named, and end up just sticking with something assuming I'd get a good suggestion. Renamed.

dnabre commented:
It really shouldn't take me so many tried to get the whitespace CI check to pass, but I'm making sure it passes. 

I committed and pushed everything to my repo, and this PR is showing all those, so I think it should be good.

dnabre commented:
Did a Close with comment, don't think that did what I thought it would. 

If there is something I need to submit to trigger review/submission of the pull request, point me to it (assuming all review/fixes are good).

BenBE commented:
Reviewing and submission is done manually. There's nothing on your side that needs to be done for this, except for pushing your up-to-date code. :)

I'll be busy the next week, thus might take a bit until I get around to doing a full review. But maybe one of the others in the team step up.

BenBE commented:
@dnabre Can you have a look at locally squashung your changes (cf. `git rebase` -> `fixup`) to make the history less cluttered? TIA.

dnabre commented:
Though I got it to work, but it had mixed in other people's commits.

BenBE commented:
> @BenBE, thought I got it to work, but it had mixed in other people's commits.
> 
> Did
> 
> ```
> git clone git@github.com:dnabre/htop.git
> cd htop
> git checkout windows
> git rebase -i 34630b6  # (has of first commit on PR)
>     -> editor
>     pick for first line
>     squash for rest
> ```

```
git clone git@github.com:dnabre/htop.git
cd htop
git checkout windows
git rebase -i --autosquash main # <- Base has to be the branch you splitting off from
```

In the editor you may (should) start off with just selecting one commit, move around its line after the commit you want to merge it with, and changing its operation to `squash` (merging, combining their commit message) or `fixup` (applying changes but discarding the commit message of the second one).

NB: You can use `git commit --fixup=<commit>` when fixing typos and such, which will automatically do the sorting for you when doing the above `git rebase` command.

> Looked here, thought it had worked, but there were other commits than mine in the log. So I reverted the commit (should have asked for help before doing that).

Generally, when unhappy with the state you are in for a rebase, you can use `git rebase --abort` to return to the state before you started that operation. Similar applies to `git merge --abort`, `git cherry-pick --abort` and possibly some more. If the operation you're in is `git pull`, depending on its mode, it's `git merge --abort` OR `git rebase --abort`. Just as a small aside. ;-)

FWIW: Looking at `git reflog` might be helpful at times too, if you accidentally screwed things up. In that case `get reset --hard <commit>` on the correct branch will likely help. ;-)

> Any guidance on what I did wrong, or which parts were correct, would be helpful. Exactly how to do the squash/merge (I under the purpose and goal at least) is the right combination of compilated and possible to do multiple ways that tutorials/forums/stackoverflow all disagree on how to do it.

Unfortunately most wisdom of the crowds™ concerning git is kinda unhelpful.

That being said, most operations related to rebasing in git become a bit more tangible if you look at them as follows:

1. A commit ultimately identifies a snapshot of all the (tracked) files in your repository.
2. Each line in the editor for the rebase tries to change the active snapshot by applying the changes done in the specified commit compared to its parent.

Thus when `git rebase -i main` gives you two commits `01234567` and `89abcdef`, then the first pick line will apply the changes that `git show 01234567` would show you and the second one will apply the ones shown in `git show 89abcdef`. So if you were to swap those two lines, now git would actually try to first apply the changes made in the second commit `89abcdef` (compared to its parent) AND then do the changes from commit `01234567` on the result it ended up with in the first step.

**Important note:** I just updated your branch `windows` using `git rebase` and force-pushing the proper changes re-applied on top of the `main` branch. Please use the following commands if you are updating your local copy of the repository to avoid accidentally merging in this fixed version with the mangled variant of the history that your local copy might still have:

```console
git fetch --all # Update all remote references
git checkout windows # Switch to correct branch
git reset --hard <your_repo>/windows # CAUTION: Forcefully overrides your local history with the one fetched from the remote
```

You should then end up on commit 3367ecfff07be7644b88c9580ce35623d1ecc9d1.

dnabre commented:
Spent most of free time today reading git stuff. `git-rebase` is complicated enough to be considered witchcraft apparently. Half the stuff out there is wrong, and the other half doesn't understand it but knows what incantations they need to get what they want accomplished. Got some good information from the hacktoberfest discord though, which helped my overall understanding of things. 

I wasn't rebasing against the `main` branch was the principle issue. No idea how I didn't get that from your comment.  My `main` branch has some of the commits applied to it, which I think is will also screw things up. I think I know how to make a branch without any of the commits that hit main, so I can do it against that. 

`git commit --fixup`  is definitely something I'm going to use a lot (now that I know about it).  Copying the whitespace check to a local alias so I can run it before commits will also help a lot. 

At any point, you can just not accept the PR, and I can resubmit the PR as a single commit easily enough. Getting right is more important to me ATM than getting it finished. 

The patience and guidance is appreciated.

dnabre commented:
Was able to make a branch of my `main` from the last commit before my additions, to give me clean reference point branch (`main_clean`). 

Had a little trouble with the rebasing squash. I tried to have everything save the last commit  as squash/fixup, and the last commit pick. Was getting the error:

```
error: cannot 'squash' without a previous commit
You can fix this with 'git rebase --edit-todo' and then run 'git rebase --continue'.
Or you can abort the rebase with 'git rebase --abort'.
```

I think my `main_clean` might have been a ref off, should the editor during rebase normally show a ref before the series I want to squash? 

Not sure if it is related, but it couldn't squash/drop the last commit.  It was a empty line removal, that I had become unnecessary by re-ordering/dropping commits. No error message, just  left me  with two commits, my squashed one and my last commit. The final commit being a PR to my own repo (not sure how I managed that). Not sure if that PR  (which had automatically been merged to my uncleaned `main`) was the issue. 

So I just  pushed the squashed commit and removed  that last commit with a `--force-with-lease` .  That may not be the most elegant way of handling that dangling commit.  I know that's in the category of commands that can break things.  I ran a lot of `--dry-runs` first, and it was giving me what I wanted/expected. 

In the end, I got it squashed into a single commit.

BenBE commented:
> Had a little trouble with the rebasing squash. I tried to have everything save the last commit as squash/fixup, and the last commit pick. Was getting the error:
> 
> ```
> error: cannot 'squash' without a previous commit
> You can fix this with 'git rebase --edit-todo' and then run 'git rebase --continue'.
> Or you can abort the rebase with 'git rebase --abort'.
> ```

You need to first pick at least one base commit. From there you can squash everything else.

> I think my `main_clean` might have been a ref off, should the editor during rebase normally show a ref before the series I want to squash?

No. The commit you are rebasing onto is not part of the list of commits you apply on top of it. It's just showing all the commits which are reachable from `HEAD`, but not from the commit you are rebasing onto.

> Not sure if it is related, but it couldn't squash/drop the last commit. It was a empty line removal, that I had become unnecessary by re-ordering/dropping commits. No error message, just left me with two commits, my squashed one and my last commit. The final commit being a PR to my own repo (not sure how I managed that). Not sure if that PR (which had automatically been merged to my uncleaned `main`) was the issue.

By default `git rebase` flattens the whole history, thus dropping any merge commits and re-arranging all the commits that led to them.

If you want to skip a commit, you can either drop it in the editor, or use the `skip` operation. An explanatory text should be below the list of operations in the rebase editor.

> So I just pushed the squashed commit and removed that last commit with a `--force-with-lease` . That may not be the most elegant way of handling that dangling commit. I know that's in the category of commands that can break things. I ran a lot of `--dry-runs` first, and it was giving me what I wanted/expected.

Actually, I run `git push --force-with-lease` so often, I got it on an alias (called `pfush`; pun on `push` and German `pfuschen`, meaning to do something badly)

> In the end, I got it squashed into a single commit.

Will take a look later.

BenBE commented on docs/winnt/cross-build-for-mingw64.md:
````diff
@@ -0,0 +1,173 @@
+```
+This configuration is based on the recommended ones for compilation to Windows from NCurses developers ( [README.MinGW](https://github.com/mirror/ncurses/blob/master/README.MinGW) with the addition of `--enable-widec` to provide Unicode support. Script `build_ncurses.sh` is provided to configure with these options.
+
+Standard `make` and `make install` to build the library and install it and its headers into `$MINGW_ROOT`
````
It seems `make install` in this step tries to unlink an existing global system installation of `ncurses`:

```
$ make install
( cd include && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/include“ wird betreten
mkdir -p /mingw-root/build-root/winnt/include/ncursesw
installing curses.h in /mingw-root/build-root/winnt/include/ncursesw
installing unctrl.h in /mingw-root/build-root/winnt/include/ncursesw
installing ncurses_dll.h in /mingw-root/build-root/winnt/include/ncursesw
installing term.h in /mingw-root/build-root/winnt/include/ncursesw
installing termcap.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./tic.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./term_entry.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./nc_tparm.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./win32_curses.h in /mingw-root/build-root/winnt/include/ncursesw
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/include“ wird verlassen
( cd ncurses && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/ncurses“ wird betreten
mkdir -p /mingw-root/build-root/winnt/lib
installing ../lib/libncursesw.a as /mingw-root/build-root/winnt/lib/libncursesw.a
/usr/bin/install -c -m 644 ../lib/libncursesw.a /mingw-root/build-root/winnt/lib/libncursesw.a
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/ncurses“ wird verlassen
( cd progs && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/progs“ wird betreten
mkdir -p /mingw-root/build-root/winnt/bin
/usr/bin/install -c -s tic.exe     /mingw-root/build-root/winnt/bin/`echo tic.exe|       sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s toe.exe     /mingw-root/build-root/winnt/bin/`echo toe.exe|       sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
linking infotocap.exe to tic.exe
( cd /mingw-root/build-root/winnt/bin && ln -s -f `echo tic.exe|       sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` `echo infotocap.exe| sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` )
linking captoinfo.exe to tic.exe
( cd /mingw-root/build-root/winnt/bin && ln -s -f `echo tic.exe|       sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` `echo captoinfo.exe| sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` )
/usr/bin/install -c -s infocmp.exe /mingw-root/build-root/winnt/bin/`echo infocmp.exe|   sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s clear.exe   /mingw-root/build-root/winnt/bin/`echo clear.exe|     sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s tabs.exe    /mingw-root/build-root/winnt/bin/`echo tabs.exe|      sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s tput.exe    /mingw-root/build-root/winnt/bin/`echo tput.exe|      sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s tset.exe    /mingw-root/build-root/winnt/bin/`echo tset.exe|      sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
linking reset.exe to tset.exe
( cd /mingw-root/build-root/winnt/bin && ln -s -f `echo tset.exe|      sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` `echo reset.exe|     sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/progs“ wird verlassen
( cd panel && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/panel“ wird betreten
installing ../lib/libpanelw.a as /mingw-root/build-root/winnt/lib/libpanelw.a
/usr/bin/install -c -m 644 ../lib/libpanelw.a /mingw-root/build-root/winnt/lib/libpanelw.a
installing ./panel.h in /mingw-root/build-root/winnt/include/ncursesw
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/panel“ wird verlassen
( cd menu && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/menu“ wird betreten
installing ../lib/libmenuw.a as /mingw-root/build-root/winnt/lib/libmenuw.a
/usr/bin/install -c -m 644 ../lib/libmenuw.a /mingw-root/build-root/winnt/lib/libmenuw.a
installing ./eti.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./menu.h in /mingw-root/build-root/winnt/include/ncursesw
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/menu“ wird verlassen
( cd form && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/form“ wird betreten
installing ../lib/libformw.a as /mingw-root/build-root/winnt/lib/libformw.a
/usr/bin/install -c -m 644 ../lib/libformw.a /mingw-root/build-root/winnt/lib/libformw.a
installing ./form.h in /mingw-root/build-root/winnt/include/ncursesw
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/form“ wird verlassen
( cd test && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/test“ wird betreten
make[1]: Für das Ziel „install“ ist nichts zu tun.
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/test“ wird verlassen
( cd misc && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/misc“ wird betreten
mkdir -p /mingw-root/build-root/winnt/share
mkdir -p /mingw-root/build-root/winnt/share/tabset
DESTDIR= \
prefix=/mingw-root/build-root/winnt \
exec_prefix=/mingw-root/build-root/winnt \
bindir=/mingw-root/build-root/winnt/bin \
top_srcdir=.. \
srcdir=. \
datadir=/mingw-root/build-root/winnt/share \
ticdir=/mingw-root/build-root/winnt/share/terminfo \
source=terminfo.tmp \
cross_compiling=yes \
/bin/sh ./run_tic.sh
** Building terminfo database, please wait...
Running /usr/bin/tic to install /mingw-root/build-root/winnt/share/terminfo ...

        You may see messages regarding extended capabilities, e.g., AX.
        These are extended terminal capabilities which are compiled
        using
                tic -x
        If you have ncurses 4.2 applications, you should read the INSTALL
        document, and install the terminfo without the -x option.

ncurses 6.3.20211021
1788 entries written to /mingw-root/build-root/winnt/share/terminfo
** built new /mingw-root/build-root/winnt/share/terminfo
** sym-linked /mingw-root/build-root/winnt/lib/terminfo for compatibility
installing std
installing stdcrt
installing vt100
installing vt300
finished install.data
/usr/bin/install -c ncurses-config /mingw-root/build-root/winnt/bin/ncursesw6-config
installing ncursesw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncursesw.pc': Permission denied
installing ncursesw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncursesw.pc': Permission denied
installing panelw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/panelw.pc': Permission denied
installing menuw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/menuw.pc': Permission denied
installing formw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/formw.pc': Permission denied
installing ncurses++w.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncurses++w.pc': Permission denied
make[1]: *** [Makefile:150: install] Fehler 1
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/misc“ wird verlassen
make: *** [Makefile:138: install] Fehler 2
```

dnabre replied:
I've run into this file testing.  It's trying to install the pkg-config files for the mingw64 version to the global system because that's where pkg-config tells it to.  pkg-config should be disabled. I'll take a look at it, could attach your config.log?

BenBE replied:
[config.log for ncurses](https://github.com/htop-dev/htop/files/9930220/config.log)

In my case it's properly installing the files in the proper platform directory (see start of the `make install` output), but further down it operating on files of the global system. With a prefix set this should not happen.

dnabre replied:
These are the errors I'm seeing:

```
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncursesw.pc': Permission denied
installing ncursesw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncursesw.pc': Permission denied
installing panelw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/panelw.pc': Permission denied
installing menuw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/menuw.pc': Permission denied
installing formw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/formw.pc': Permission denied
installing ncurses++w.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncurses++w.pc': Permission denied
make[1]: *** [Makefile:150: install] Fehler 1
```

This is trying to install the pkg-config files for mingw64, but it's getting the location to put those files from running `pkg-config` instead basing it on the prefix. Getting pkg-config setup properly just seems unnecessary. Looking at your `config.log`  the wrong options are being given to `configure` . It should be disabling pkg-config stuff from being built (and then installed).  e.g., `--enable-pc-files` should be the opposite. 

Cleaning up the ncurses options along with testing that it was building binaries that would run properly on Windows was the last thing I did before my git adventure. I think I might have dropped a commit with the final version of the flags. I'll look at it and fix it.

dnabre replied:
Was able to take a quick look. The `build_ncurses.sh` script and the configure options in the instructions are out of sync, with the script having the correct ones (from a quick look).

The scripts have been handy when I've been working, but would it be best to remove them?  If only to remove the possibility of them being out of sync with the documentation?

BenBE replied:
I only looked at the instructions earlier when I tested this.

Given that documentation should usually not contain executables I'd suggest to remove those scripts.

BenBE commented on docs/winnt/cross-build-for-mingw64.md:
````diff
@@ -0,0 +1,173 @@
+```
+./autogen.sh
+
+# Make sure configure can find the cross0built ncurses library and headers
````
```suggestion
# Make sure configure can find the cross-built ncurses library and headers
```

BenBE commented on docs/winnt/cross-build-for-mingw64.md:
```diff
@@ -0,0 +1,173 @@
+export CFLAGS="-I$MINGW_ROOT/include"
+export LDFLAGS="-L$MINGW_ROOT/lib"
+
+./configure \
+	--host=$HOST_PLATFORM \
+	--prefix=$MINGW_ROOT \
+	--enable-static \
+	--enable-unicode
```
This fails to find `sys/utsname.h`, as that one is listed in `configure.ac` to be required. Instead when targeting platform `winnt` this should be skipped in `configure.ac`.

Workaround:

```diff
diff --git a/configure.ac b/configure.ac
index 53163366..9893f500 100644
--- a/configure.ac
+++ b/configure.ac
@@ -138,7 +138,6 @@ AC_CHECK_HEADERS([ \
     strings.h \
     sys/param.h \
     sys/time.h \
-    sys/utsname.h \
     unistd.h
    ], [], [AC_MSG_ERROR([can not find required generic header files])])
 
@@ -160,6 +159,10 @@ if test "$my_htop_platform" = darwin; then
    AC_CHECK_HEADERS([mach/mach_time.h])
 fi
 
+if test "$my_htop_platform" != unsupported; then
+   AC_CHECK_HEADERS([sys/utsname.h], [], [AC_MSG_ERROR([can not find required generic header files])])
+fi
+
 # ----------------------------------------------------------------------
 
 
```

Ideally this should be handled as a separate `winnt` platform, but this is upcoming work.

ookiineko replied:
> Workaround:

FYI gnulib has a stub for this, See: https://www.gnu.org/software/gnulib/manual/html_node/sys_002futsname_002eh.html and https://github.com/coreutils/gnulib/blob/f36daeac96c30aa5ed42ffa9ccd8cd4366b08b13/lib/uname.c#L56

dnabre replied:
> > Workaround:
> 
> FYI gnulib has a stub header and implementation for this on MinGW/MSVC, See: https://www.gnu.org/software/gnulib/manual/html_node/sys_002futsname_002eh.html and https://github.com/coreutils/gnulib/blob/f36daeac96c30aa5ed42ffa9ccd8cd4366b08b13/lib/uname.c#L56

The `winnt` build shouldn't be using `utsname.h` (it's functionality isn't available on Windows). We want its contents to be missing so that builds that include break ASAP.

BenBE commented on docs/winnt/cross-build-for-mingw64.md:
````diff
@@ -0,0 +1,173 @@
+	--enable-unicode
+```
+
+Then the standard `make` and `make install` to build and put the files into the $MINGW_ROOT tree.
````
This fails to compile in several places due to some headers implicitly used:
- `langinfo.h`
- `pwd.h`
- `sys/resource.h`
- `sys/select.h`
- `sys/wait.h`

Furthermore some defines/symbols are not present:
- `AT_FDCWD`
- `AT_SYMLINK_NOFOLLOW`
- `lstat`
- `readlink`

Overall this list seems manageable.

BenBE commented:
@dnabre Any updates from your side regarding the cleanup of this PR?

dnabre commented:
Pretty sure the document is done in terms of cleanups and such.

Actually doing the port has dropped off my project list though, so not sure if there's any reason to add them to the repo.

BenBE commented:
> Pretty sure the document is done in terms of cleanups and such.

AFAICS there are still some remarks re style/consistency open. As well as some comments regarding issues I noticed while going through the documentation and trying to replicate things. Please have a look at the review comments above and mark things that are done as resolved. TIA.

> Actually doing the port has dropped off my project list though, so not sure if there's any reason to add them to the repo.

NP. I think keeping this PR as a starting point for cross-compilation docs suffices, as the actual porting will be a bit more work than just running things through the compiler. Having a clean starting point for such future work would be worthwhile nonetheless.

TIA for the work so far.

dnabre commented:
> AFAICS there are still some remarks re style/consistency open. As well as some comments regarding issues I noticed while going through the documentation and trying to replicate things. Please have a look at the review comments above and mark things that are done as resolved. TIA.

I thought cleanup stuff was complete, but that's just going by memory, not looking at the history


> NP. I think keeping this PR as a starting point for cross-compilation docs suffices, as the actual porting will be a bit more work than just running things through the compiler. Having a clean starting point for such future work would be worthwhile nonetheless.

I'll try to find some time and go back through the reviews/commits and finish cleaning up the instructions at least.

dnabre commented:
Looks like @ookiineko have it up and running using a Windows-runners + MSYS2 with MSYS2 dependent build (same for Cygwin) in short order.  Switching the MSYS2 build from being MSYS2 dependent to native Windows will be just changing one line in the setup. The Linux-based process code will have to be changed, of course, but the build setup will be there. 

I'm going through a doing a final clean up/proof on this. Minor updates for things that have changed. Having this rundown for doing it all from Linux won't hurt, especially since it's already put together, though not sure it's shorted path at this point for fully-native Windows port.

ookiineko commented:
> Switching the MSYS2 build from being MSYS2 dependent to native Windows will be just changing one line in the setup.

a quick note, you also need to change the depended package names

for example: autotools -> mingw-w64-x86_64-autotools; ncurses-devel -> mingw-w64-x86_64-ncurses

the unprefixed packages are for Msys2 runtime. in addition to mingw, you can also use other environment (including different arches like clangarm64), the list of those environments and differences between them: https://www.msys2.org/docs/environments/ , about package prefixes, see: https://www.msys2.org/docs/package-naming/

good luck with the native port!

dnabre commented:
My htop fork is very behind, and the tree got pretty messed up at some point. Likely the cause of the CI  build-ubuntu-latest{-full-feature...,-gcc-static}. Ended up touch up all the commands in this, since my test setup got lost  (a pair of mirrored SSD dying together sucks). The end result is a lot cleaner in terms of being able to actually try to cross-build the fully `configured`ed code. 

Need to do a proofread and touch up need time I've got some spare energy for everything from "Configure and Cross-compile htop" down.  Anything above detail-level that should be included/removed  topic-wise?

ookiineko commented on docs/winnt/cross-build-for-mingw64.md:
```diff
@@ -0,0 +1,184 @@
+# Cross-compiling htop for Windows from Linux
+
+An intermediate build using Cygwin/MSYS is being worked on by @ookiineko as of 2023-09-28 (see #1097)
```
while you can use them as a starting point for the native port, I think Cygwin/Msys ports are NOT intermediate builds, they should be considered as standalone unix-like platforms. See: https://github.com/htop-dev/htop/issues/1097#issuecomment-1740194685

dnabre replied:
I think their existence should be mentioned in this document.  Not sure on the best wording.

I don't agree with your view of the platforms or usefulness them beyond a native port, but that have nothing to do with this PR.

BenBE commented on docs/winnt/cross-build-for-mingw64.md:
```diff
@@ -22,19 +22,18 @@ and it will use the appropriate compiler, linkers, and libraries for that platfo
 ## Setting up the Cross-Build Environment
 
-Going through specifics, we going assume a base install of Debian Linux 11.x, bullseye, `x86_64`. Because it has (almost) everything needed in its packaging system, and because I have setup where I can  easily spin up a clean install of it. While package names may differ,  most major distributions will have same tools available.
+Going through specifics, we going assume a base install of Ubuntu 22.04 for `x86_64`. The same version used by GitHub's ubuntu-latest runner. It has (almost) everything needed in its packaging system. While package names may differ,  most major distributions will have same tools available.
```
```suggestion
Going through specifics, we are going to assume a base install of Ubuntu 22.04 for `x86_64`; the same version also used by GitHub's `ubuntu-latest` runner. It has (almost) everything needed in its packaging system. While package names may differ, most major distributions will have these tools available.
```

BenBE commented on docs/winnt/cross-build-for-mingw64.md:
```diff
@@ -114,12 +108,12 @@ Clone the `htop` repo and checkout the `windows` branch (currently work on the `
 ./autogen.sh
 
-# Make sure configure can find the cross0built ncurses library and headers
+# Make sure configure can find the cross built NCurses library and headers
```
```suggestion
# Make sure configure can find the cross-built NCurses library and headers
```

dnabre commented:
Don't even remember if I intended to close this. 
Got a workflow to remove silly errors (extra spaces, minor grammar, etc), The cross-compile tools setup and build is now complete error free on my testing. 

If this document is still useful,  submitting it as a new pull request from a clean repo (my fork) would really simplify the amount of pain github is giving me.

dnabre commented:
Didn't mean to close

dnabre commented:
Ok, branch should probably be renamed from 'windows' to 'winnt'. Doing so at the moment messing up a bunch of stuff, so I switched the name back. 

Updated/fixed all current errors. Got some extensions for VS Code that greatly help my proofreading. markdownlint does a great job of finding whitespace errors and making everything consistent.
